dynamiqs/smesolve/smesolve.py

```diff
@@ -1,23 +1,105 @@
 from __future__ import annotations
 
-from typing import Any
-
-import numpy as np
+import torch
 from torch import Tensor
 
+from ..options import Euler, Options
+from ..utils.tensor_formatter import TensorFormatter
+from ..utils.tensor_types import OperatorLike, TDOperatorLike, TensorLike
+from .euler import SMEEuler
+
 
 def smesolve(
-    H: Tensor,
-    Ls: list[Tensor],
-    Ms: list[Tensor],
-    etas: list[float],
-    rho0: Tensor,
-    tlist: np.ndarray,
+    H: TDOperatorLike,
+    jump_ops: OperatorLike | list[OperatorLike],
+    rho0: OperatorLike,
+    t_save: TensorLike,
+    etas: TensorLike,
     ntrajs: int,
-    solver: str,
-    options: dict[str, Any] | None = None,
-) -> Tensor:
-    raise NotImplementedError(
-        'the Stochastic Master Equation solvers are not implemented yet, come '
-        'back soon!'
+    *,
+    t_meas: TensorLike | None = None,
+    seed: int | None = None,
+    exp_ops: OperatorLike | list[OperatorLike] | None = None,
+    options: Options | None = None,
+) -> tuple[Tensor, Tensor | None, Tensor | None]:
+    # H: (b_H?, n, n), rho0: (b_rho0?, n, n) -> (y_save, exp_save, meas_save) with
+    #    - y_save: (b_H?, b_rho0?, ntrajs, len(t_save), n, n)
+    #    - exp_save: (b_H?, b_rho0?, ntrajs, len(exp_ops), len(t_save))
+    #    - meas_save: (b_H?, b_rho0?, ntrajs, len(meas_ops), len(t_meas) - 1)
+
+    # check jump_ops
+    if isinstance(jump_ops, list) and len(jump_ops) == 0:
+        raise ValueError(
+            'Argument `jump_ops` must be a non-empty list of tensors. Otherwise,'
+            ' consider using `ssesolve`.'
+        )
+
+    # format and batch all tensors
+    formatter = TensorFormatter(options.dtype, options.device)
+    H_batched, rho0_batched = formatter.format_H_and_state(H, rho0, state_to_dm=True)
+    H_batched = H_batched.unsqueeze(2)
+    rho0_batched = rho0_batched.unsqueeze(2).repeat(1, 1, ntrajs, 1, 1)
+    # H_batched: (b_H, 1, 1, n, n)
+    # rho0_batched: (b_H, b_rho0, ntrajs, n, n)
```

Comment:
Having comments in the middle kinda breaks the flow when reading (it looks like 2 different blocks of code). You can put these at the top below `# format and batch all tensors` (line 37)